File: src/main/java/catchla/yep/view/holder/DribbbleShotViewHolder.java
package catchla.yep.view.holder;

import android.graphics.Bitmap;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;

import com.nostra13.universalimageloader.core.assist.FailReason;
import com.nostra13.universalimageloader.core.listener.ImageLoadingListener;

import java.util.List;

import catchla.yep.R;
import catchla.yep.adapter.DribbbleShotsAdapter;
import catchla.yep.model.DribbbleShot;
import catchla.yep.model.DribbbleShotImage;
import catchla.yep.util.ImageLoaderWrapper;

/**
 * Created by mariotaku on 15/6/3.
 */
public class DribbbleShotViewHolder extends RecyclerView.ViewHolder implements ImageLoadingListener {

    private final DribbbleShotsAdapter adapter;

    private final ImageView imageView;
    private final View imageProgress;

    public DribbbleShotViewHolder(final DribbbleShotsAdapter adapter, final View itemView) {
        super(itemView);
        this.adapter = adapter;
        imageView = (ImageView) itemView.findViewById(R.id.image_view);
        imageProgress = itemView.findViewById(R.id.image_progress);
    }

    public void displayShot(DribbbleShot shot) {
        final DribbbleShotImage image = getBestImage(shot.getImages());
        final ImageLoaderWrapper imageLoader = adapter.getImageLoader();
        if (image != null) {
            imageLoader.displayImage(image.getUrl(), imageView, this, null);
        } else {
            imageLoader.cancelDisplayTask(imageView);
        }
    }

    private DribbbleShotImage getBestImage(final List<DribbbleShotImage> images) {
        if (images == null) return null;
        for (DribbbleShotImage image : images) {
            if ("normal".equals(image.getResolution())) return image;
        }
        return null;
    }

    @Override
    public void onLoadingStarted(final String imageUri, final View view) {
        imageProgress.setVisibility(View.VISIBLE);
    }

    @Override
    public void onLoadingFailed(final String imageUri, final View view, final FailReason failReason) {
        imageProgress.setVisibility(View.GONE);
    }

    @Override
    public void onLoadingComplete(final String imageUri, final View view, final Bitmap loadedImage) {
        imageProgress.setVisibility(View.GONE);
    }

    @Override
    public void onLoadingCancelled(final String imageUri, final View view) {
        imageProgress.setVisibility(View.GONE);
    }
}

File: src/main/java/catchla/yep/view/holder/DribbbleShotViewHolder.kt
package catchla.yep.view.holder

import android.graphics.Bitmap
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.ImageView

import com.nostra13.universalimageloader.core.assist.FailReason
import com.nostra13.universalimageloader.core.listener.ImageLoadingListener

import catchla.yep.R
import catchla.yep.adapter.DribbbleShotsAdapter
import catchla.yep.model.DribbbleShot
import catchla.yep.model.DribbbleShotImage
import catchla.yep.util.ImageLoaderWrapper

/**
 * Created by mariotaku on 15/6/3.
 */
class DribbbleShotViewHolder(private val adapter: DribbbleShotsAdapter, itemView: View) : RecyclerView.ViewHolder(itemView), ImageLoadingListener {

    private val imageView: ImageView
    private val imageProgress: View

    init {
        imageView = itemView.findViewById(R.id.image_view) as ImageView
        imageProgress = itemView.findViewById(R.id.image_progress)
    }

    fun displayShot(shot: DribbbleShot) {
        val image = getBestImage(shot.images)
        val imageLoader = adapter.imageLoader
        if (image != null) {
            imageLoader.displayImage(image.url, imageView, this, null)
        } else {
            imageLoader.cancelDisplayTask(imageView)
        }
    }

    private fun getBestImage(images: List<DribbbleShotImage>?): DribbbleShotImage? {
        if (images == null) return null
        for (image in images) {
            if ("normal" == image.resolution) return image
        }
        return null
    }

    override fun onLoadingStarted(imageUri: String, view: View) {
        imageProgress.visibility = View.VISIBLE
    }

    override fun onLoadingFailed(imageUri: String, view: View, failReason: FailReason) {
        imageProgress.visibility = View.GONE
    }

    override fun onLoadingComplete(imageUri: String, view: View, loadedImage: Bitmap) {
        imageProgress.visibility = View.GONE
    }

    override fun onLoadingCancelled(imageUri: String, view: View) {
        imageProgress.visibility = View.GONE
    }
}
